src/buf/fixed/plumbing/pool.rs

```diff
@@ -0,0 +1,180 @@
+                *state = BufState::CheckedOut;
+                (init_len, next)
+            }
+            BufState::CheckedOut => panic!("buffer is checked out"),
+        };
+
+        // Update the head of the free list for this capacity.
+        match next {
+            Some(i) => {
+                *free_head = i;
+            }
+            None => {
+                self.free_buf_head_by_cap.remove(&cap);
+            }
+        }
+
+        // Safety: the allocated array under the pointer is valid
+        // for the lifetime of self, a free buffer index is inside the array,
+        // as also asserted by the indexing operation on the states array
+        // that has the same length.
+        let iovec = unsafe { self.raw_bufs.as_ptr().add(index).read() };
+        debug_assert_eq!(iovec.iov_len, cap);
+        Some(CheckedOutBuf {
+            iovec,
+            init_len,
+            index: index as u16,
+        })
+    }
+
+    pub(crate) fn notify_on_next(&mut self, cap: usize) -> Arc<Notify> {
+        let notify = self.notify_next_by_cap.entry(cap).or_default();
+        Arc::clone(notify)
+    }
+
+    fn check_in_internal(&mut self, index: u16, init_len: usize) {
+        let cap = self.iovecs()[index as usize].iov_len;
+        let state = &mut self.states[index as usize];
+        debug_assert!(
+            matches!(state, BufState::CheckedOut),
+            "the buffer must be checked out"
+        );
+
+        // Link the buffer as the new head of the free list for its capacity.
+        // Recently checked in buffers will be first to be reused,
+        // improving cache locality.
+        let next = self.free_buf_head_by_cap.insert(cap, index);
+
+        *state = BufState::Free { init_len, next };
+
+        if let Some(notify) = self.notify_next_by_cap.get(&cap) {
+            // Wake up a single task pending on `next`
+            notify.notify_one();
+        }
+    }
+}
+
+impl FixedBuffers for Pool {
+    fn iovecs(&self) -> &[iovec] {
+        // Safety: the raw_bufs pointer is valid for the lifetime of self,
+        // the length of the states array is also the length of buffers array
+        // by construction.
+        unsafe { slice::from_raw_parts(self.raw_bufs.as_ptr(), self.states.len()) }
+    }
+
+    unsafe fn check_in(&mut self, index: u16, init_len: usize) {
+        self.check_in_internal(index, init_len)
+    }
+}
+
+impl Drop for Pool {
+    fn drop(&mut self) {
+        let iovecs = unsafe {
+            Vec::from_raw_parts(self.raw_bufs.as_ptr(), self.states.len(), self.orig_cap)
+        };
+        for (i, iovec) in iovecs.iter().enumerate() {
+            match self.states[i] {
+                BufState::Free { init_len, next: _ } => {
+                    let ptr = iovec.iov_base as *mut u8;
+                    let cap = iovec.iov_len;
+                    let v = unsafe { Vec::from_raw_parts(ptr, init_len, cap) };
+                    mem::drop(v);
```

Comment:
Each buffer must be individually allocated since it's supplied to the constructor as a `Vec`.
I don't think io-uring requires the buffers to be aligned; tests pass with buffers of size 80, for example.

I put a `mem::drop` call in there to illustrate that the `Vec` is recovered from raw pointer and size specifically to be dropped and thus to get the buffer properly deallocated.

Reply:
I'm not sure about this. If the constructor didn't do the individual allocators, it doesn't seem right that the drop would assume individual deallocations.

Reply:
At least some kernel ops require page aligned buffers. I don't recall which right now but why shouldn't the interface allow the caller to decide for themselves how the allocations are done? Does that make the API too onerous?

Reply:
> If the constructor didn't do the individual allocators, it doesn't seem right that the drop would assume individual deallocations.

The constructor takes ownership of a sequence of `Vec` that have to be allocated using the global allocator, so drop deallocates these `Vec` objects. If the application wants a more precise control over the allocations, perhaps it should consider its own implementation of a `FixedBuffers` collection (when we get around to make the trait public).